src/drivers/pmw3901/pmw3901.cpp

```diff
@@ -0,0 +1,944 @@
+
+	report.sensor_id = 0;
+	report.quality = 255;
+
+	if (_optical_flow_pub == nullptr) {
+
+		_optical_flow_pub = orb_advertise(ORB_ID(optical_flow), &report);
+
+	} else {
+
+		orb_publish(ORB_ID(optical_flow), _optical_flow_pub, &report);
+	}
+
+	/* post a report to the ring */
+	_reports->force(&report);
+
+	/* notify anyone waiting for data */
+	poll_notify(POLLIN);
+
+	ret = OK;
+
+	perf_end(_sample_perf);
+	return ret;
+
+}
+
+
+int
+PMW3901::readMotionCount(int16_t &deltaX, int16_t &deltaY)
+{
+	int ret;
+
+	uint8_t data[10] = { DIR_READ(0x02), 0, DIR_READ(0x03), 0, DIR_READ(0x04), 0,
+			     DIR_READ(0x05), 0, DIR_READ(0x06), 0
+			   };
+
+	ret = transfer(&data[0], &data[0], 10);
+
+	if (OK != ret) {
+		perf_count(_comms_errors);
+		DEVICE_LOG("spi::transfer returned %d", ret);
+		return ret;
+	}
+
+	deltaX = ((int16_t)data[5] << 8) | data[3];
+	deltaY = ((int16_t)data[9] << 8) | data[7];
+
+	ret = OK;
+
+	return ret;
+
+}
+
+
+void
+PMW3901::start()
+{
+	/* reset the report ring and state machine */
+	_reports->flush();
+
+	/* schedule a cycle to start things */
+	work_queue(LPWORK, &_work, (worker_t)&PMW3901::cycle_trampoline, this, 1000);
+
+	/* notify about state change */
+	struct subsystem_info_s info = {};
+	info.present = true;
+	info.enabled = true;
+	info.ok = true;
+	info.subsystem_type = subsystem_info_s::SUBSYSTEM_TYPE_OPTICALFLOW;
+
+	static orb_advert_t pub = nullptr;
+
+	if (pub != nullptr) {
+		orb_publish(ORB_ID(subsystem_info), pub, &info);
+
+	} else {
+		pub = orb_advertise(ORB_ID(subsystem_info), &info);
+	}
+}
+
+void
+PMW3901::stop()
+{
+	work_cancel(HPWORK, &_work);
+}
+
+void
+PMW3901::cycle_trampoline(void *arg)
+{
+	PMW3901 *dev = (PMW3901 *)arg;
+
+	dev->cycle();
+}
+
+void
+PMW3901::cycle()
+{
+	collect();
+
+	/* schedule a fresh cycle call when the measurement is done */
+	work_queue(LPWORK,
+		   &_work,
+		   (worker_t)&PMW3901::cycle_trampoline,
+		   this,
+		   USEC2TICK(PMW3901_SAMPLE_RATE));
+
+}
+
+void
+PMW3901::print_info()
+{
+	perf_print_counter(_sample_perf);
+	perf_print_counter(_comms_errors);
+	printf("poll interval:  %u ticks\n", _measure_ticks);
+	_reports->print_info("report queue");
+}
+
+
+/**
+ * Local functions in support of the shell command.
+ */
+namespace pmw3901
+{
+
+PMW3901	*g_dev;
+
+void	start(uint8_t rotation);
+void	stop();
+void	test();
+void	reset();
+void	info();
+
+
+/**
+ * Start the driver.
+ */
+void
+start(uint8_t rotation)
+{
+	int fd;
+
+	if (g_dev != nullptr) {
+		errx(1, "already started");
+	}
+
+	/* create the driver */
+	g_dev = new PMW3901(rotation, PMW3901_BUS);
+
+	if (g_dev == nullptr) {
+		goto fail;
+	}
+
+	if (OK != g_dev->init()) {
+		goto fail;
+	}
+
+	/* set the poll rate to default, starts automatic data collection */
+	fd = open(PMW3901_DEVICE_PATH, O_RDONLY);
+
+	if (fd < 0) {
+		goto fail;
+	}
+
+	if (ioctl(fd, SENSORIOCSPOLLRATE, SENSOR_POLLRATE_DEFAULT) < 0) {
+		goto fail;
+	}
+
+	exit(0);
```

Comment:
In general I would suggest assuming that possibility nearly everywhere as we go and even bother including them in the posix builds (sitl https://github.com/PX4/Firmware/blob/master/cmake/configs/posix_sitl_default.cmake#L6, rpi, etc).

At a minimum it will cause the code to be checked by tools like coverity and clang-tidy.

All that's actually required to "port" most NuttX drivers to Linux/QuRT is to use the px4 headers instead of anything NuttX specific, and prevent it from directly exiting (exit, err).